avr-hal-generic/src/eeprom.rs

```diff
@@ -0,0 +1,339 @@
+                            helper::set_erase_mode(self);
+                            self.eecr.write(|w| w.eepe().set_bit()); // Start Erase-only operation.
+                        }
+                    }
+                    //Now we know that _no_ bits need to be erased to '1'.
+                    else {
+                        // Check if any bits are changed from '1' in the old value.
+                        if diff_mask != 0 {
+                            // Now we know that _some_ bits need to the programmed to '0'.
+                            self.eedr.write(|w| w.bits(data)); // Set EEPROM data register.
+                            helper::set_write_mode(self);
+                            self.eecr.write(|w| w.eepe().set_bit()); // Start Write-only operation.
+                        }
+                    }
+                }
+            }
+
+            fn raw_erase_byte(&mut self, address: u16) {
+                unsafe {
+                    helper::set_address(self, address as $addrwidth);
+                    // Now we know that all bits should be erased.
+                    helper::set_erase_mode(self);
+                    // Start Erase-only operation.
+                    self.eecr.write(|w| w.eepe().set_bit());
+                }
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! impl_eeprom_attiny {
+    (
+        hal: $HAL:ty,
+        peripheral: $EEPROM:ty,
+        capacity: $capacity:literal,
+        addr_width: $addrwidth:ty,
+        set_address: |$periph_var:ident, $address:ident| $set_address:block,
+    ) => {
+        mod helper {
+            #[inline]
+            pub unsafe fn wait_read(regs: &$EEPROM) {
+                while regs.eecr.read().eepe().bit_is_set() {}
+            }
+            #[inline]
+            pub unsafe fn set_address(regs: &$EEPROM, address: $addrwidth) {
+                wait_read(regs);
+                let $periph_var = regs;
+                let $address = address;
+                $set_address
+            }
+            #[inline]
+            pub unsafe fn set_erasewrite_mode(regs: &$EEPROM) {
+                regs.eecr.write(|w| {
+                    // Set Master Write Enable bit...and and Erase+Write mode mode..
+                    w.eempe().set_bit().eepm().atomic()
+                });
+            }
+            #[inline]
+            pub unsafe fn set_erase_mode(regs: &$EEPROM) {
+                regs.eecr.write(|w| {
+                    // Set Master Write Enable bit, and Erase-only mode..
+                    w.eempe().set_bit().eepm().erase()
+                });
+            }
+            #[inline]
+            pub unsafe fn set_write_mode(regs: &$EEPROM) {
+                regs.eecr.write(|w| {
+                    // Set Master Write Enable bit, and Write-only mode..
+                    w.eempe().set_bit().eepm().write()
+                });
+            }
+        }
+        impl $crate::eeprom::EepromOps<$HAL> for $EEPROM {
```

Comment:
I liked your original approach of having a third macro for the common code so it isn't duplicated between the ATmega and ATtiny macros.  How about moving the `impl EepromOps for $EEPROM` into a `impl_eeprom_common!{}` macro?

Reply:
review fixed